### PR TITLE
Enable 2-tx reconfiguration with CFT

### PR DESCRIPTION
**Pull request by wintersteiger**
Mainly passing an additional option through the various parts of the infrastructure.

The new `reconfiguration_test_cft_2tx` should probably be disabled or moved to the long/daily tests for now.

### CODE REVIEW

**Comment by jumaffre** [comment]
This will need a new entry in the changelog too.

File: CHANGELOG.md
Comment (jumaffre) [inline CHANGELOG.md:10]:
```suggestion
### Added 

- Added support for 2-transaction reconfiguration with CFT consensus (#3097), see [documentation](https://microsoft.github.io/CCF/main/overview/consensus/bft.html#two-transaction-reconfiguration).
```

File: src/node/node_state.h
Comment (jumaffre) [inline src/node/node_state.h:959]: This could be making use of `.value_or()` to be more concise.

File: src/node/rpc/node_frontend.h
Comment (jumaffre) [inline src/node/rpc/node_frontend.h:371]: This seems to be called twice in the `/node/join` endpoint. Could we only call it once and pass the `reconfiguration_type` result around instead?

File: src/node/rpc/node_frontend.h
Comment (jumaffre) [inline src/node/rpc/node_frontend.h:459]: Do we really need this? AFAICS, this is for a new node that's waiting to be trusted by members but is retired in the meantime. So it might have been briefly trusted but never participated in the consensus (otherwise it would have already joined). So maybe joining nodes should never join if they're retiring?

File: src/node/rpc/node_frontend.h
Comment (jumaffre) [inline src/node/rpc/node_frontend.h:1362]: I think the consensus checks underneath are already handled by the framework (see `frontend.h`).

File: src/node/rpc/node_frontend.h
Comment (jumaffre) [inline src/node/rpc/node_frontend.h:1403]: This isn't about resharing, is it?

File: tests/infra/consortium.py
Comment (jumaffre) [inline tests/infra/consortium.py:298]: From a UX perspective, it seems a little awkward to change the check for successfully retire node proposal logic depending on the type of reconfiguration. In all cases, the node should become `RETIRED` after a relatively short timeout, sometimes passing through the `RETIRING` state?

File: tests/infra/network.py
Comment (jumaffre) [inline tests/infra/network.py:641]: There's already an argument for this, `--ledger-recovery-timeout`. Perhaps we want to rename it and use it here?

File: tests/reconfiguration.py
Comment (jumaffre) [inline tests/reconfiguration.py:450]: As discussed, can we move this to `partititons_test.py` as this requires `sudo`.

Comment (jumaffre) [reply tests/reconfiguration.py:450]: Also, I don't understand why the partition of f+1 backups happens so early. This could be the reason why `new_node` cannot join the service (no commit over its join transaction). I believe it would be more pertinent to only isolate `new_node` from the rest of the service before `trust_node(new_node)` and then check that progress on the service doesn't depend on this node (e.g. by partitioning some other backups). Then, when all partitions are lifted, the node should join, observe its addition, orc and become part of consensus.

File: src/consensus/aft/raft.h
Comment (jumaffre) [inline src/consensus/aft/raft.h:565]: Since, the `hostname` and `port` for a given node should always stay the same, isn't this just `conf1 == conf2`?

File: src/consensus/aft/raft.h
Comment (jumaffre) [inline src/consensus/aft/raft.h:661]: Isn't this path also used in CFT, when 2tx reconfiguration is enabled?

File: src/consensus/aft/orc_requests.h
Comment (jumaffre) [inline src/consensus/aft/orc_requests.h:92]: As discussed, if the primary is being retired, the delay should be a function of the election timeout so that ORC aren't submitted when we know there's little certainty that a new primary has been elected.

File: src/consensus/aft/orc_requests.h
Comment (jumaffre) [inline src/consensus/aft/orc_requests.h:43]: It took me a while to wrap my head around this: this is actually sending a request to itself and relying on the forwarding mechanism in the generic frontend to send the request to the new primary. I think this is OK. However, the logic in `http_node_client.h` does not actually wait for the response to the request (it will take a while to receive it from the forwardee) and I think we should wait for this asynchronously, and also wait for commit on this transaction.

File: src/consensus/aft/orc_requests.h
Comment (jumaffre) [inline src/consensus/aft/orc_requests.h:69]: Why is there a limit here? If the node never sends the ORC, will the node be stuck forever?

File: src/node/rpc/node_frontend.h
Comment (wintersteiger) [reply src/node/rpc/node_frontend.h:459]: I added this because it happens during the reconfiguration tests. A node attempts to join, but is retired before it hears about it and thus can't decide whether it should submit the join RPC. So, I think it's necessary, but maybe the behaviour should be the same as for `NodeStatus::PENDING`?

File: src/node/rpc/node_frontend.h
Comment (wintersteiger) [reply src/node/rpc/node_frontend.h:1362]: I added that because it's also done in other handlers in this file, e.g. `is_primary` and `get_primary`. So, I concluded that I can't rely on the consensus being set at all time. Would this particular handler not be called at all if there's no consensus object?

File: src/node/rpc/node_frontend.h
Comment (wintersteiger) [reply src/node/rpc/node_frontend.h:1403]: Nope, good catch, that's a copy&paste gone wrong :)

File: tests/infra/consortium.py
Comment (wintersteiger) [reply tests/infra/consortium.py:298]: Not in the test scripts - if there are bugs, they may not retire at all, so we need a check with timeout. Also, the other test scripts rely on the fact that this happens pretty much instantaneously and try to re-use the ip/port of the retiring node, so we have to make sure the node is actually gone before we continue (or add waiting loops in the corresponding tests).

File: src/consensus/aft/raft.h
Comment (wintersteiger) [reply src/consensus/aft/raft.h:565]: Yes, except there's no `operator==` for `NodeInfo`. I just wanted to compare the node ids, so I opted not to add an operator. Do you think it would be better to add one?

File: src/consensus/aft/raft.h
Comment (wintersteiger) [reply src/consensus/aft/raft.h:661]: Yes, except when 1tx reconfiguration is used `conf` is never the same twice. With 2-tx reconfig this happens when the only change to a configuration is a node transitioning from retiring to retired (because of the update to the nodes table).

File: src/consensus/aft/orc_requests.h
Comment (wintersteiger) [reply src/consensus/aft/orc_requests.h:69]: Yes, it would be stuck. The limit is `SIZE_MAX` though. In the beginning I wasn't sure whether we need/want a limit, but you're right, we should keep trying forever.

File: src/consensus/aft/orc_requests.h
Comment (wintersteiger) [reply src/consensus/aft/orc_requests.h:43]: In general that is true (and I've started adding code for that in the split identity PR), but for ORCs it's not necessary to wait. The "reply" is essential the primary changing the node status in the nodes table. I believe, if necessary, all nodes will hit the same reconfiguration commit check in `raft.h` again and re-schedule new ORCs.

File: tests/reconfiguration.py
Comment (wintersteiger) [reply tests/reconfiguration.py:450]: I added the `CONFIGURATION` option in cmake scripts, so this one also runs with `sudo`.

Not sure I understand. `trust_node` runs right after the partition is introduced, I just added a `check_does_not_progress` in between for good measure. What would you have done differently?

File: tests/infra/network.py
Comment (wintersteiger) [reply tests/infra/network.py:641]: That's great, thanks! Maybe `ledger-read-timeout`?

File: src/node/rpc/node_frontend.h
Comment (wintersteiger) [reply src/node/rpc/node_frontend.h:456]: Move checks to function `is_taking_part_in_acking`. Add comment to explain why?

File: src/node/rpc/node_frontend.h
Comment (jumaffre) [reply src/node/rpc/node_frontend.h:1364]: I believe the two endpoints you mentioned are very much consensus-specific and execute outside consensus (`.set_execute_outside_consensus(ccf::endpoints::ExecuteOutsideConsensus::Locally)`), hence the extra checks. In the ORC case, these can be skipped.

File: tests/infra/consortium.py
Comment (jumaffre) [reply tests/infra/consortium.py:298]: OK. In this case, can the node already be in `retired` state in this RPC, even with a 2tx reconfiguration (i.e. if node was fast enough to retire itself)? It feels like the logic could be unified: wait for the node to be marked as `RETIRED` within a timeout. If we're in 1tx reconfiguration, the check should pass straight away.

File: src/consensus/aft/raft.h
Comment (jumaffre) [reply src/consensus/aft/raft.h:565]: `raft.h` is already chunky so defining the `operator==` for `NodeInfo` is probably simpler.

File: CMakeLists.txt
Comment (achamayou) [inline CMakeLists.txt:523]: I am not sure this rename is good, but it's definitely incomplete (uses of the flag have not been updated), and it does not belong in this PR in any case.

File: src/consensus/aft/raft.h
Comment (jumaffre) [reply src/consensus/aft/raft.h:661]: As per `hooks.h`, I thought that retiring nodes were marked as `std::nullopt` in the configuration, whereas retired ones were explicitly passed in the `retirees` list. 

In any case, I'm surprised this calls into BFT code (`get_bft_offset`) even though this is on the CFT path.

File: src/consensus/aft/orc_requests.h
Comment (achamayou) [inline src/consensus/aft/orc_requests.h:69]: why 250?

File: src/consensus/aft/raft.h
Comment (achamayou) [inline src/consensus/aft/raft.h:626]: Would suggest "retiring nodes" since "retiring" is a state name.

File: src/consensus/aft/raft.h
Comment (achamayou) [inline src/consensus/aft/raft.h:3256]: s/retirees/retiring_nodes/

**Comment by achamayou** [comment]
@wintersteiger this needs to either be disabled in release, or made safe when used against a 1tx network (ie. refuse to join with a helpful error) before merging.

File: src/consensus/aft/orc_requests.h
Comment (wintersteiger) [reply src/consensus/aft/orc_requests.h:69]: Because it felt right! This is a re-schedule of a failed RPC (need to keep sending ORCs forever in general), but I didn't want to reschedule without any delay to avoid liveness issues.

File: src/consensus/aft/raft.h
Comment (wintersteiger) [reply src/consensus/aft/raft.h:661]: The offset is there to allow BFT to run with non-consecutive node IDs; after a view-change, the next primary is the one at view-change-number + offset in the list of nodes. This has nothing to do with this PR though.

File: src/node/rpc/node_frontend.h
Comment (wintersteiger) [reply src/node/rpc/node_frontend.h:1364]: That makes a lot of sense, thanks for pointing that out!

File: src/consensus/aft/orc_requests.h
Comment (achamayou) [reply src/consensus/aft/orc_requests.h:69]: Ok! Can we have a constexpr ORC_RPC_RETRY_INTERVAL_MS, and use that here? I think it'd be more readable.

File: src/consensus/aft/raft.h
Comment (achamayou) [reply src/consensus/aft/raft.h:643]: Ok, as discussed, I think we want to remove this, but you are absolutely right that it's nothing to do with this change.

File: .azure-pipelines-templates/matrix.yml
Comment (wintersteiger) [inline .azure-pipelines-templates/matrix.yml:28]: @achamayou Is this over-write of the new option in release mode fine here, or should I add `...=ON` to each of the other configurations?

Comment (achamayou) [reply .azure-pipelines-templates/matrix.yml:28]: This is fine, imo.

**Review by achamayou — APPROVED** [review]

File: CHANGELOG.md
Comment (jumaffre) [inline CHANGELOG.md:11]: The documentation should be moved out of the BFT page.

File: src/consensus/aft/raft.h
Comment (jumaffre) [inline src/consensus/aft/raft.h:3442]: You mentioned in our discussion that we stop scheduling the ORC when we see our `RETIRED` state as committed. It seems that the only condition to stop scheduling the ORC is when an ORC has been forwarded to a node (`submit_orc()` returns true in `orc_requests.h`), which isn't enough to guarantee that the ORC has successfully been applied (the request may have been lost, or the transaction rolled back). What am I missing?

File: src/kv/test/stub_consensus.h
Comment (jumaffre) [inline src/kv/test/stub_consensus.h:152]: `learner_nodes` for consistency?

File: src/node/rpc/frontend.h
Comment (jumaffre) [inline src/node/rpc/frontend.h:113]: Is it worth adding this to the changelog? And also checking that this is the return status we get from the ORC when the node is retiring?

File: src/consensus/aft/raft.h
Comment (achamayou) [reply src/consensus/aft/raft.h:3442]: We do need to keep retrying until we've observed the ORC committed or our transition complete (retired or trusted), a node having accepted (to forward) the RPC isn't enough.

Comment (wintersteiger) [reply src/consensus/aft/raft.h:3442]: It will keep scheduling new ORCs on every `commit()` _until_ it becomes `RETIRED` (which also means that the entire rest of the new configuration is ready to be applied, so no more ORCs needed). But, we don't need to know whether the ORC RPC call itself succeeded - we can't do anything to correct the error anyways, all we can do is retry. Am I missing something?

File: src/node/rpc/node_frontend.h
Comment (jumaffre) [inline src/node/rpc/node_frontend.h:1386]: I'm a little worried about the call to `consensus->orc()` on line 1383. This endpoint writes to the KV and can thus be aborted or rolled back. However, the endpoint also has some side effects on the internal state of the consensus (`orc_sets`) which does not seem to be transactional in nature (e.g. I do not see any rollback logic for it). 

I think `orc_sets` should be recorded in the KV store instead (e.g. whether it's passed the quorum or not could be recorded as a `bool` in the `configuration` table), which then passed to the consensus via a hook (which might not be necessary as we already have a hook on the configuration table).

File: tests/infra/consortium.py
Comment (jumaffre) [inline tests/infra/consortium.py:331]: AIUI, the test would pass if the node is fast enough to transition from `LEARNER` to `TRUSTED` before we check its state in the store. Wouldn't it be equivalent to `self.wait_for_node_to_exist_in_store(remote_node, node_to_retire.node_id, timeout, NodeStatus.TRUSTED)` for both 1tx and 2tx reconfiguration?

Comment (wintersteiger) [reply tests/infra/consortium.py:331]: Yes, I think that's right, I just didn't want to run the waiting loop here as it's fine if the node is still a learner.

File: src/node/rpc/frontend.h
Comment (jumaffre) [reply src/node/rpc/frontend.h:113]: As discussed, the ORC client doesn't have to check the response and should keep retrying until it's seen itself as `TRUSTED` from the consensus.

File: src/node/rpc/node_frontend.h
Comment (jumaffre) [reply src/node/rpc/node_frontend.h:1386]: As discussed, this should be OK as long as i) nodes retry to send ORCs until they've seen their `TRUSTED`ness committed, ii) `orc_sets` (and other consensus state updated in `consensus->orc()` are flushed when a node becomes candidate. We should also try to add a test for this (end-to-end or unit test?) but it's OK to merge this PR as it is as long as the existing reconfiguration test is stable.

File: src/consensus/aft/raft.h
Comment (jumaffre) [reply src/consensus/aft/raft.h:3442]: https://github.com/microsoft/CCF/pull/3097#discussion_r741986320

File: src/consensus/aft/raft.h
Comment (achamayou) [inline src/consensus/aft/raft.h:740]:
```suggestion
    // guaranteed that we will eventually receive all of them, since all
```

File: src/node/rpc/node_frontend.h
Comment (wintersteiger) [reply src/node/rpc/node_frontend.h:1386]: I created #3157 for this and assigned myself, so I'll resolve the conversation here.

**Review by jumaffre — APPROVED** [review]

**Pull request merged**
